shell/platform/common/cpp/client_wrapper/include/flutter/encodable_value.h

```diff
@@ -6,16 +6,150 @@
 #define FLUTTER_SHELL_PLATFORM_COMMON_CPP_CLIENT_WRAPPER_INCLUDE_FLUTTER_ENCODABLE_VALUE_H_
 
 #include <assert.h>
+
 #include <cstdint>
 #include <map>
 #include <string>
 #include <utility>
+#include <variant>
 #include <vector>
 
 namespace flutter {
 
 static_assert(sizeof(double) == 8, "EncodableValue requires a 64-bit double");
 
+// Defining USE_LEGACY_ENCODABLE_VALUE will use the original EncodableValue
+// implementation. This is a temporary measure to minimize the impact of the
+// breaking change; it will be removed in the future. If you set this, you
+// should update your code as soon as possible to use the new std::variant
+// version, or it will break when the legacy version is removed.
+#ifndef USE_LEGACY_ENCODABLE_VALUE
+
+class EncodableValue;
+
+// Convenience type aliases.
+using EncodableList = std::vector<EncodableValue>;
+using EncodableMap = std::map<EncodableValue, EncodableValue>;
+
+// An object that can contain any value or collection type supported by
+// Flutter's standard method codec.
+//
+// For details, see:
+// https://api.flutter.dev/flutter/services/StandardMessageCodec-class.html
+//
+// As an example, the following Dart structure:
+//   {
+//     'flag': true,
+//     'name': 'Thing',
+//     'values': [1, 2.0, 4],
+//   }
+// would correspond to:
+//   EncodableValue(EncodableMap{
+//       {EncodableValue("flag"), EncodableValue(true)},
+//       {EncodableValue("name"), EncodableValue("Thing")},
+//       {EncodableValue("values"), EncodableValue(EncodableList{
+//                                      EncodableValue(1),
+//                                      EncodableValue(2.0),
+//                                      EncodableValue(4),
+//                                  })},
+//   })
+//
+// The primary API surface for this object is std::variant. For instance,
+// getting a string value from an EncodableValue, with type checking:
+//   if (std::get<std::string>(value)) {
+//     std::string some_string = std::get<std::string>(value);
+//   }
+//
+// The order/indexes of the variant types is part of the API surface, and is
+// guaranteed not to change.
+class EncodableValue : public std::variant<std::monostate,
```

Comment:
The variant definition is long and repeated. How about just calling it EncodableValueVariant up top and using it once?

Reply:
I went back and forth on that; my concern is that if it's declared with a name someone might use it directly, probably on accident (e.g., picking `EncodableValueVariant` instead of `EncodableValue` in IDE autocompletion). I felt like duplicating a value that I don't expect to change (except one more time when I make this extensible for custom StandardCodec types) was better than adding something undesired to the global namespace.

I guess the other option would be to declare it in a `namespace internal {}` or something along those lines.